metrics/fds_linux.go

```diff
@@ -4,10 +4,106 @@
 
 package metrics
 
-import "os"
+import (
+	"fmt"
+	"log"
+	"syscall"
+	"unsafe"
+)
 
 func currentFDs() int {
-	// TODO(bradfitz): do this without so many allocations on Linux.
-	ents, _ := os.ReadDir("/proc/self/fd")
-	return len(ents)
+	fd, err := openProcSelfFD()
+	if err != nil {
+		return 0
+	}
+	defer syscall.Close(fd)
+
+	count := 0
+
+	const blockSize = 8 << 10
+	buf := make([]byte, blockSize) // stack-allocated; doesn't escape
+	bufp := 0                      // starting read position in buf
+	nbuf := 0                      // end valid data in buf
+	dirent := &syscall.Dirent{}
+	for {
+		if bufp >= nbuf {
+			bufp = 0
+			nbuf, err = readDirent(fd, buf)
+			if err != nil {
+				log.Printf("currentFDs: readDirent: %v", err)
+				return 0
+			}
+			if nbuf <= 0 {
+				return count
+			}
+		}
+		consumed, name := parseDirEnt(dirent, buf[bufp:nbuf])
+		bufp += consumed
+		if len(name) == 0 || string(name) == "." || string(name) == ".." {
+			continue
+		}
+		count++
+	}
+}
+
+func direntNamlen(dirent *syscall.Dirent) uint64 {
```

Comment:
Weird for a len of something held in memory to be a uint64 and not an int.

Reply:
done